src/formats/prometheus_text_format.erl

```diff
@@ -57,16 +57,52 @@ format() ->
 ?DOC("Formats `Registry` using the latest text format.").
 -spec format(Registry :: prometheus_registry:registry()) -> binary().
 format(Registry) ->
-    {ok, Fd} = ram_file:open("", [write, read, binary]),
-    Callback = fun(_, Collector) ->
-        registry_collect_callback(Fd, Registry, Collector)
-    end,
-    prometheus_registry:collect(Registry, Callback),
-    file:write(Fd, "\n"),
-    {ok, Size} = ram_file:get_size(Fd),
-    {ok, Str} = file:pread(Fd, 0, Size),
-    ok = file:close(Fd),
-    Str.
+    format_into(Registry, fun format_into_binary/2, <<>>).
+
+format_into_binary(Acc, Data) ->
+    <<Acc/binary, Data/binary>>.
+
+?DOC("""
+Formats `Registry` using the latest text format, passing the binary data for
+each collector to the format function.
+""").
+-spec format_into(
+    Registry :: prometheus_registry:registry(), fun((term(), binary()) -> term()), term()
+) -> term().
+format_into(Registry, Fmt, State) ->
+    State1 = lists:foldl(
+        format_into_collector_fn(Registry, Fmt), State, prometheus_registry:collectors(Registry)
+    ),
+    Fmt(State1, <<"\n">>).
+
+format_into_collector_fn(Registry, Fmt) ->
+    fun(Collector, Acc) ->
+        put(?MODULE, Acc),
+        prometheus_collector:collect_mf(
+            Registry, Collector, format_into_create_mf_callback_fn(Fmt)
+        ),
+        erase(?MODULE)
+    end.
+
+format_into_create_mf_callback_fn(Fmt) ->
+    fun(#'MetricFamily'{name = Name0, help = Help, type = Type, metric = Metrics}) ->
+        %% eagerly convert the name to a binary so we can copy more efficiently
+        %% in `render_metrics/3`
+        Name = iolist_to_binary(Name0),
+        Prologue = <<
+            "# TYPE ",
+            Name/binary,
+            " ",
+            (string_type(Type))/binary,
+            "\n# HELP ",
+            Name/binary,
+            " ",
+            (escape_metric_help(Help))/binary,
+            "\n"
+        >>,
+        Bin = render_metrics(Prologue, Name, Metrics),
+        put(?MODULE, Fmt(Bin, erase(?MODULE)))
```

Comment:
Bah, I made a mistake here with the order of the arguments. The function should take the state as the first argument and then the new data as the second argument. It doesn't end up making a difference for `format/1` because it just changes the order that the metrics families are formatted in - it's just concatenating the wrong way. But using `format_into/3` with a custom formatting function doesn't work properly. I'll send a follow-up PR